src/index/IndexImpl.h

```diff
@@ -399,11 +399,12 @@ class IndexImpl {
       const TripleComponent& col0String,
       std::optional<std::reference_wrapper<const TripleComponent>> col1String,
       const Permutation::Enum& permutation,
-      ad_utility::SharedConcurrentTimeoutTimer timer = nullptr) const;
+      std::shared_ptr<ad_utility::AbortionHandle> abortionHandle) const;
```

Comment:
Alternatively, as this is used so often:
Do we need a `shared_reference` type, that has the semantics of `never nullptr` but is otherwise shared?
I think that is something that can be rather useful in the case of asynchronous programming where ofthen `shared_xxx` is a reasonable compromise between safety and performance.

Reply:
If it can't be nullptr it can't be moveable either, so that's a conceptual problem there.

Reply:
I think we have discussed several possible implementations
(for example add two flags, one for "was moved" and for "was used" so that then the destructor has a check that from  "was moved" follows "not was used". That way we have a write on every access (maybe a mutable atomic) but no branch. Altenratviely we could assert on each access that it's not null.
But you are right: In c++ with non-destructive moves there is no way to let the type system handle that statically.  I know that this is a non-issue in Rust.

I think this can however be a separate PR, are you interested in implementing this, or should I do it myself or delegate it?

Reply:
> I think this can however be a separate PR, are you interested in implementing this, or should I do it myself or delegate it?

No, not interested at all :D 
It seems like a nice thing to have in general, but not something that's necessarily interesting to implement